packages/core-sdk/lib/Connectors/BaseConnector.ts

```diff
@@ -1,27 +1,64 @@
 import { ConnectorEvent, IConnectorAdapters } from "../Types";
 import EventEmitter from "eventemitter3";
-import { IConnector } from "../Types/IConnector";
+import { IConnector, OnNetworkChangeOptions } from "../Types/IConnector";
 import { Callback } from "../Utils/Typings";
 import { IBlockchainConfig } from "../Types/IBlockchainConfig";
 import { IConnectorMetadata } from "../Types";
+import { blockchainConfigMap } from "../Blockchains";
 
 export abstract class BaseConnector implements IConnector {
   protected emitter = new EventEmitter<ConnectorEvent>();
   _adapter: IConnectorAdapters | undefined;
 
   constructor(
     public readonly metadata: IConnectorMetadata,
-    public readonly config: IBlockchainConfig
+    public config: IBlockchainConfig
   ) {}
 
-  protected abstract _connect(): Promise<IConnectorAdapters>;
+  protected abstract _connect(
+    config?: IBlockchainConfig
+  ): Promise<IConnectorAdapters>;
 
-  async connect(): Promise<IConnectorAdapters> {
+  async connect(
+    { tryReconnection }: OnNetworkChangeOptions = {
+      tryReconnection: false,
+    }
+  ): Promise<IConnectorAdapters> {
     if (this.adapter) {
       return this.adapter;
     }
-    const adapters = await this._connect();
+
+    const adapters = await this._connect(this.config);
     this.adapter = adapters;
+    this.emitter.emit("Connected", adapters);
+
+    if (tryReconnection) {
+      this.on("NetworkChanged", async (chainId) => {
+        const config = Object.values(blockchainConfigMap).find(
+          (cfg) => cfg.chainId === chainId
+        );
+        if (config === undefined) {
+          throw new Error("Target blockchain is not supported.");
+        }
+        const adapters = await this.changeNetwork(config);
+        this.adapter = adapters;
+        this.emitter.emit("Connected", adapters);
+      });
+    }
+
+    return adapters;
```

Comment:
When using core-sdk and calling the connect method we are awaiting to the response. 
If we pass the option to try the network change we are expecting that it will try to get the correct adapter on the correct blockchain before getting the result.

The way its currently written will return the adapter on the current blockchain despite being the desired chain or not, and will change it silently after resolving the connect method causing uncontrollable behaviour.

We would need to do the following:
1. Connect as you are already doing on line 31.
2. Check if the chain is correct.
3. If the chain is correct we return the adapters and set this.adapters = adapters

It would be the original behaviour, but if the chain id is diferent and the tryReconnection is set to true, we would need to return a Promise that will resolve after the first network change and return its adapter whenever they are.